rainier-core/src/main/scala/com/stripe/rainier/core/RandomVariable.scala

```diff
@@ -226,4 +238,85 @@ object RandomVariable {
   def fit[L, T](l: L, seq: Seq[T])(
       implicit toLH: ToLikelihood[L, T]): RandomVariable[Unit] =
     toLH(l).fit(seq)
+
+  private[core] def waic(samples: List[Array[Double]],
+                         variables: List[Variable],
+                         target: Target): Double =
+    if (target.nRows == 0)
+      0.0
+    else {
+      val cf = Compiler.default.compile(
+        target.placeholderVariables ++ variables,
+        List(("logLikelihood", target.real)))
+      val globalBuf = new Array[Double](cf.numGlobals)
+      val inputBuf = new Array[Double](cf.numInputs)
+      val data = target.placeholderVariables.map { v =>
+        target.placeholders(v)
+      }.toArray
+      val sampleArray = samples.toArray
+      var i = 0
```

Comment:
trying to be allocation-free from this point on...

Reply:
This should probably be a comment and the exception a TODO comment